R/sum.R

```diff
@@ -231,27 +231,56 @@ calculateBinaryOperation <- function(x, y,
         if (!is.null(previous.counts <- attr(x, "n.sum")))
         {
             counts.to.sum <- list(previous.counts, (!is.na(input[[2L]])) * 1L)
-            dimensions <- lapply(input, dim)
+            count.names <- lapply(counts.to.sum, dimnames)
+            dimensions <- lapply(counts.to.sum, standardizedDimensions)
             dimensions.equal <- identical(dimensions[[1L]], dimensions[[2L]])
-            matching.required <- if (dimensions.equal) "No" else "Yes - hide unmatched"
-            counts.to.sum <- matchDimensionElements(counts.to.sum, match.rows = matching.required,
-                                                    match.columns = matching.required, warn = FALSE,
-                                                    function.name = function.name)
-            current.counts <- operation(counts.to.sum[[1L]], counts.to.sum[[2L]])
+            if (!dimensions.equal)
+                counts.to.sum[[1L]] <- subsetFirstInputToMatchSecondInput(counts.to.sum[[1L]], input[[1L]])
+            current.counts <- counts.to.sum[[1L]] + counts.to.sum[[2L]]
+            previous.counts <- counts.to.sum[[1L]]
         } else
         {
             non.missing.vals <- lapply(input, function(x) (!(is.na(x))) * 1L)
-            current.counts <- operation(non.missing.vals[[1L]], non.missing.vals[[2L]])
+            current.counts <- `+`(non.missing.vals[[1L]], non.missing.vals[[2L]])
         }
     }
-    if (is.primitive(operation))
+    with.mean.attribute <- as.character(substitute(operation)) == ".updateVarianceWithMissingOptions"
+    if (with.mean.attribute)
     {
+        previous.mean <- attr(x, "mean")
+        no.previous.mean <- is.null(previous.mean)
+        if (remove.missing && no.previous.mean)
+            mean.input <- lapply(input, removeMissing)
+        else if (remove.missing)
+            mean.input <- list(input[[1L]], removeMissing(input[[2L]]))
+        else
+            mean.input <- input
+        if (no.previous.mean)
+            current.mean <- (mean.input[[1L]] + mean.input[[2L]])/current.counts
+        else
+        {
+            if (!dimensions.equal)
+                previous.mean <- subsetFirstInputToMatchSecondInput(previous.mean, mean.input[[1L]])
+            current.mean <- previous.mean * previous.counts/current.counts + mean.input[[2L]]/current.counts
+
+            if (any(update.mean <- is.nan(current.mean) & !is.na(input[[2L]])))
+                current.mean[update.mean] <- input[[2L]][update.mean]
+        }
+        if (!is.null(single.obs <- attr(x, "only.single.obs")))
+        {
+            if (!dimensions.equal)
+                single.obs <- subsetFirstInputToMatchSecondInput(single.obs, input[[1L]])
+            attr(input[[1L]], "only.single.obs") <- single.obs
+        }
+        attr(input[[1L]], "mean") <- previous.mean
+        attr(input[[1L]], "n.sum") <- previous.counts
+    } else
         input <- if (remove.missing) lapply(input, removeMissing) else input
```

Comment:
Since `removeMissing` replaces `NA` values with 0, this change can break `Min` and `Max` with multiple inputs. e.g. `Min(c(NA, NA, 2), c(NA, 1, 1))` returns `0,0,1`.

Reply:
Hmm, ok. I'll change and add tests for `Max` and `Min` in this case too.

Reply:
Resolved in 516bb63